include/swift/SIL/SILInstructionWorklist.h

```diff
@@ -22,94 +22,312 @@
-  }
+  void erase(SILInstruction *instruction) { worklist.erase(instruction); }
 
   /// Remove the top element from the worklist.
-  SILInstruction *removeOne() {
-    SILInstruction *instruction = worklist.pop_back_val();
-    worklistMap.erase(instruction);
-    return instruction;
+  SILInstruction *pop_back_val() {
+    return worklist.pop_back_val().getValueOr(nullptr);
   }
 
   /// When an instruction has been simplified, add all of its users to the
   /// worklist, since additional simplifications of its users may have been
   /// exposed.
-  void addUsersToWorklist(ValueBase *instruction) {
-    for (auto *use : instruction->getUses())
-      add(use->getUser());
-  }
-
   void addUsersToWorklist(SILValue value) {
-    for (auto *use : value->getUses())
+    for (auto *use : value->getUses()) {
       add(use->getUser());
+    }
   }
 
   /// When an instruction has been simplified, add all of its users to the
   /// worklist, since additional simplifications of its users may have been
   /// exposed.
   void addUsersOfAllResultsToWorklist(SILInstruction *instruction) {
     for (auto result : instruction->getResults()) {
-      addUsersToWorklist(result);
+      addUsersToWorklist(&*result);
+    }
+  }
+
+  /// Check that the worklist is empty and nuke the backing store if it is
+  /// large.
+  void resetChecked() {
+    assert(worklist.empty() && "Vector empty, but the map is not?");
+
+    // Do an explicit clear, shrinking the storage if needed.
+    worklist.clear();
+  }
+
+  /// Find usages of \p instruction and replace them with usages of \p result.
+  ///
+  /// Intended to be called during visitation after \p instruction has been
+  /// removed from the worklist.
+  ///
+  /// \p instruction the instruction whose usages will be replaced
+  /// \p result the instruction whose usages will replace \p instruction
+  ///
+  /// \return whether the instruction was deleted or modified.
+  bool replaceInstructionWithInstruction(SILInstruction *instruction,
+                                         SILInstruction *result
+#ifndef NDNEBUG
+                                         ,
+                                         std::string instructionDescription
+#endif
+  ) {
+    if (result != instruction) {
+      assert(&*std::prev(SILBasicBlock::iterator(instruction)) == result &&
+             "Expected new instruction inserted before existing instruction!");
+
+      withDebugStream([&](llvm::raw_ostream &stream, StringRef loggingName) {
+        stream << loggingName << ": Old = " << *instruction << '\n'
+               << "  "
+               << "  New = " << *result << '\n';
+      });
+
+      // Everything uses the new instruction now.
+      replaceInstUsesPairwiseWith(instruction, result);
+
+      // Push the new instruction and any users onto the worklist.
+      add(result);
+      addUsersOfAllResultsToWorklist(result);
+
+      eraseInstFromFunction(*instruction);
+
+      return true;
+    } else {
```

Comment:
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

Feel free to do in a subsequent commit.

Reply:
https://github.com/apple/swift/pull/27062